### PR TITLE
Feat/arc ingest command handler support

### PR DESCRIPTION
**Pull request by cain4chain**
## Description of Changes
feat(arc_ingest.go) : add main arc-ingest method
feat(jsonutil.go) : add new JSON util method to parse byte array and handle errors
feat(limited_body_reader.go) : Add specified utility method to read data chunks, will be used in other command handlers such as submit_transaction.go

## Linked Issues / Tickets
Closes : Add arc-ingest command handler support #17

## Testing Procedure
feat(arc_ingest_test.go) : Add happy test and table driven tests for validation errors
feat(limited_body_reader_test) : Test cases for testing limited_body_reader.go method.

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have run the linter

### CODE REVIEW

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	bb, err := hex.DecodeString(r.TxID)
+	if err != nil {
+		return chainhash.Hash{}, errors.Join(err, ErrInvalidTxIDFormat)
+	}
+	if len(bb) != chainhash.HashSize {
+		return chainhash.Hash{}, ErrInvalidTxIDLength
+	}
+	var hash chainhash.Hash
+	copy(hash[:], bb)
+	return hash, nil
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]: You could use a function from `go-sdk`: [NewHashFromHex](https://github.com/bsv-blockchain/go-sdk/blob/dbbb2cde494d1852eb6e312b286aa9ef975bbbf0/chainhash/hash.go#L110)
- It does necessary validation also.

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+		case errors.Is(err, ErrInvalidMerklePathFormat):
+			jsonutil.SendHTTPResponse(w, http.StatusBadRequest, NewFailureArcIngestHandlerResponse(ErrInvalidMerklePathFormat.Error()))
+			return
+		case errors.Is(err, ErrInvalidTxIDFormat):
+			jsonutil.SendHTTPResponse(w, http.StatusBadRequest, NewFailureArcIngestHandlerResponse(ErrInvalidTxIDFormat.Error()))
+			return
+		default:
+			jsonutil.SendHTTPResponse(w, http.StatusBadRequest, NewFailureArcIngestHandlerResponse(err.Error()))
+			return
+		}
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]: This could be moved to private helper function.

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	if err != nil {
+		return chainhash.Hash{}, nil, err
+	}
+	return txIDHash, merklePath, nil
+}
+
+// Handle processes an ARC ingest request, handling the validation, converting
+// the request data into the correct format, and invoking the provider to handle
+// the new Merkle proof. It also manages timeout and error responses.
+func (h *ArcIngestHandler) Handle(w http.ResponseWriter, r *http.Request) {
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go:177]: Maybe, such methods should receive a `ctx context.Context` as a first argument. 
- And correctly react when provided context is "canceled" (because of timeout, for example)

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	}
+	errCh := make(chan error, 1)
+	defer close(errCh)
+	go func() { errCh <- h.provider.HandleNewMerkleProof(r.Context(), &txIDHash, merklePath) }()
+	select {
+	case err := <-errCh:
+		if err == nil {
+			jsonutil.SendHTTPResponse(w, http.StatusOK, NewSuccessArcIngestHandlerResponse())
+			return
+		}
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]: I suppose, it could be made by [context.WithTimeout ](https://golang.cafe/blog/golang-context-with-timeout-example.html).

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+// NewArcIngestHandler creates a new instance of an ArcIngestHandler, utilizing
+// the provided Merkle proof provider and any optional configurations.
+func NewArcIngestHandler(provider NewMerkleProofProvider, opts ...ArcIngestHandlerOption) (*ArcIngestHandler, error) {
+	if provider == nil {
+		return nil, fmt.Errorf("provider is nil")
+	}
+	h := ArcIngestHandler{
+		provider:         provider,
+		requestBodyLimit: jsonutil.RequestBodyLimit1GB,
+		responseTimeout:  10 * time.Second,
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]: Please, make a constant for it.

File: pkg/server/internal/app/jsonutil/limited_body_reader.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+// It will read chunks of data from the body and accumulate them until the read limit is reached
+// or until the end of the body is encountered. If the read limit is exceeded, it returns an error.
+// If the reading or writing to the buffer fails, an error is returned.
+//
+// The function reads the data in chunks of 64KB (defined by the buffer size),
+// ensuring that no more than the allowed `ReadLimit` bytes are read.
+func (l *LimitedBodyReader) Read() ([]byte, error) {
+	reader := io.LimitReader(l.Body, l.ReadLimit+1)
+	buff := bytes.NewBuffer(nil)
+	bb := make([]byte, 64*1024)
```
Comment (chris-4chain) [inline pkg/server/internal/app/jsonutil/limited_body_reader.go]: It looks like "magic number" - please make a constant with descriptive name for it. 
- e.g. `chunkSize`

File: pkg/server/internal/app/commands/arc_ingest.go (hunk 3 above)
Comment (mgosek-4chain) [reply pkg/server/internal/app/commands/arc_ingest.go:177]: It would be better to create a decorator instead of duplicating code across each handler's codebase. However, we can add that "catch" in a separate PR. Sounds good? @chris-4chain

File: pkg/server/internal/app/commands/arc_ingest.go (hunk 4 above)
Comment (cain4chain) [reply pkg/server/internal/app/commands/arc_ingest.go]: Ok thanks I will update this.

File: pkg/server/internal/app/commands/arc_ingest_test.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	"net/http/httptest"
+	"testing"
+
+	"github.com/4chain-ag/go-overlay-services/pkg/server/internal/app/commands"
+	"github.com/4chain-ag/go-overlay-services/pkg/server/internal/app/commands/testutil"
+	"github.com/4chain-ag/go-overlay-services/pkg/server/internal/app/jsonutil"
+	"github.com/stretchr/testify/require"
+)
+
+// Test_ArcIngestHandler_ShouldRespondsWith200AndCallsProvider tests the successful handling of a valid request
```
Comment (mgosek-4chain) [inline pkg/server/internal/app/commands/arc_ingest_test.go]: According to our convention, we don't add comments to unit tests, so please remove them :)

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+	// ErrMissingRequiredRequestFieldsDefinition is returned when the request body is missing
+	// required fields, such as the transaction ID and Merkle path.
+	ErrMissingRequiredRequestFieldsDefinition = errors.New("missing required fields: txid, merkle path")
+	// ErrMissingRequiredTxIDFieldDefinition is returned when the request body is missing
+	// the required transaction ID field.
+	ErrMissingRequiredTxIDFieldDefinition = errors.New("missing required field: txid")
+	// ErrMissingRequiredMerklePathFieldDefinition is returned when the request body is missing
+	// the required Merkle path field.
+	ErrMissingRequiredMerklePathFieldDefinition = errors.New("missing required field: merkle path")
+)
```
Comment (mgosek-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]:
```
var (
	// ErrInvalidTxIDFormat is returned when the transaction ID is not in a valid format (e.g., not hexadecimal).
	ErrInvalidTxIDFormat = errors.New("invalid transaction ID format")

	// ErrInvalidTxIDLength is returned when the transaction ID does not match the expected length (typically 64 characters for a SHA-256 hash).
	ErrInvalidTxIDLength = errors.New("invalid transaction ID length")

	// ErrInvalidMerklePathFormat is returned when the Merkle path is malformed or does not conform to the expected structure.
	ErrInvalidMerklePathFormat = errors.New("invalid Merkle path format")

	// ErrMissingRequiredRequestFieldsDefinition is returned when the request body is missing
	// required fields, such as the transaction ID and Merkle path.
	ErrMissingRequiredRequestFieldsDefinition = errors.New("missing required fields: txid, merkle path")

	// ErrMissingRequiredTxIDFieldDefinition is returned when the request body is missing
	// the required transaction ID field.

	ErrMissingRequiredTxIDFieldDefinition = errors.New("missing required field: txid")

	// ErrMissingRequiredMerklePathFieldDefinition is returned when the request body is missing
	// the required Merkle path field.
	ErrMissingRequiredMerklePathFieldDefinition = errors.New("missing required field: merkle path")
)
```

Added missing spaces.

File: pkg/server/internal/app/commands/arc_ingest.go (hunk 4 above)
Comment (mgosek-4chain) [reply pkg/server/internal/app/commands/arc_ingest.go]: @cain4chain 

```
	ctx, cancel := context.WithTimeout(r.Context(), h.responseTimeout)
	defer cancel()

	err = h.provider.HandleNewMerkleProof(ctx, &txIDHash, merklePath)
	switch {
	case errors.Is(err, context.DeadlineExceeded):
		jsonutil.SendHTTPResponse(w, http.StatusGatewayTimeout, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusGatewayTimeout)))

	case errors.Is(err, context.Canceled):
		jsonutil.SendHTTPResponse(w, http.StatusRequestTimeout, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusRequestTimeout)))

	case err != nil:
		jsonutil.SendHTTPResponse(w, http.StatusInternalServerError, NewFailureArcIngestHandlerResponse(http.StatusText(http.StatusInternalServerError)))

	default:
		jsonutil.SendHTTPResponse(w, http.StatusOK, NewSuccessArcIngestHandlerResponse())
	}
```

Additionally, it would be good to add some structured logging and extract it into a separate function.

File: pkg/server/internal/app/commands/arc_ingest_test.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+			require.Equal(t, tc.expectedStatus, res.StatusCode)
+
+			var response commands.ArcIngestHandlerResponse
+			err = jsonutil.DecodeResponseBody(res, &response)
+			require.NoError(t, err)
+
+			require.Contains(t, response.Message, tc.expectedError.Error())
+			require.Equal(t, "error", response.Status)
+		})
+	}
```
Comment (mgosek-4chain) [inline pkg/server/internal/app/commands/arc_ingest_test.go]:
```
	tests := map[string]struct {
		method             string
		payload            commands.ArcIngestRequest
		expectedResponse   commands.ArcIngestHandlerResponse
		expectedHTTPStatus int
	}{
		"should fail with 405 when HTTP method is GET": {
			method: http.MethodGet,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidHTTPMethod.Error()),
			expectedHTTPStatus: http.StatusMethodNotAllowed,
		},

		"should fail with 400 when all required fields are missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "",
				MerklePath:  "",
				BlockHeight: 0,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredRequestFieldsDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID field is missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredTxIDFieldDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when MerklePath field is missing": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  "",
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrMissingRequiredMerklePathFieldDefinition.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID format is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "invalid-hex-string",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidTxIDFormat.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when TxID length is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        "1234",
				MerklePath:  testutil.NewValidTestMerklePath(t),
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidTxIDLength.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
		"should fail with 400 when MerklePath format is invalid": {
			method: http.MethodPost,
			payload: commands.ArcIngestRequest{
				TxID:        testutil.ValidTxId,
				MerklePath:  "invalid-merkle-path",
				BlockHeight: 848372,
			},
			expectedResponse:   commands.NewFailureArcIngestHandlerResponse(commands.ErrInvalidMerklePathFormat.Error()),
			expectedHTTPStatus: http.StatusBadRequest,
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			// given:
			mock := testutil.NewMerkleProofProviderMock(nil, tc.payload.BlockHeight)
			handler, err := commands.NewArcIngestHandler(mock)
			require.NoError(t, err)

			ts := httptest.NewServer(http.HandlerFunc(handler.Handle))
			defer ts.Close()

			req, err := http.NewRequest(tc.method, ts.URL, testutil.RequestBody(t, tc.payload))
			require.NoError(t, err)

			// when:
			res, err := ts.Client().Do(req)
			require.NoError(t, err)
			defer res.Body.Close()

			// then:
			require.Equal(t, tc.expectedHTTPStatus, res.StatusCode)

			var actualResponse commands.ArcIngestHandlerResponse
			err = jsonutil.DecodeResponseBody(res, &actualResponse)
			require.NoError(t, err)

			require.Equal(t, tc.expectedResponse, actualResponse)
		})
	}
```

File: pkg/server/internal/app/commands/arc_ingest_test.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+		"should fail with 405 when HTTP method is PUT": {
+			method: http.MethodPut,
+			payload: commands.ArcIngestRequest{
+				TxID:        testutil.ValidTxId,
+				MerklePath:  testutil.NewValidTestMerklePath(t),
+				BlockHeight: 848372,
+			},
+			expectedStatus: http.StatusMethodNotAllowed,
+			expectedError:  commands.ErrInvalidHTTPMethod,
+		},
```
Comment (mgosek-4chain) [inline pkg/server/internal/app/commands/arc_ingest_test.go]: One of them is sufficient; we don't need both testing the same edge case :)

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+// sending appropriate HTTP responses based on the specific error type.
+func (h *ArcIngestHandler) handleMerkleProofResult(w http.ResponseWriter, err error) {
+	switch {
+	case errors.Is(err, context.DeadlineExceeded):
+		slog.Error(fmt.Sprintf("[ArcIngest] Merkle proof processing timed out: %v", err))
+		jsonutil.SendHTTPResponse(w, http.StatusGatewayTimeout, NewFailureArcIngestHandlerResponse(ErrMerkleProofProcessingTimeout.Error()))
+
+	case errors.Is(err, context.Canceled):
+		slog.Error(fmt.Sprintf("[ArcIngest] Merkle proof processing canceled: %v", err))
+		jsonutil.SendHTTPResponse(w, http.StatusRequestTimeout, NewFailureArcIngestHandlerResponse(ErrMerkleProofProcessingCanceled.Error()))
```
Comment (mgosek-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go:206]: Minor: could you please add unit tests to cover both timeout and cancellation scenarios?

File: pkg/server/internal/app/commands/arc_ingest.go [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+// appropriate format, and forwarding the data to the overlay engine for processing.
+type ArcIngestHandler struct {
+	provider         NewMerkleProofProvider
+	requestBodyLimit int64
+	responseTimeout  time.Duration
+}
+
+// decode reads and decodes the request body into the provided destination
+// struct. It ensures the request body size is within the allowed limit.
+func (h *ArcIngestHandler) decode(body io.Reader, dst *ArcIngestRequest) error {
```
Comment (chris-4chain) [inline pkg/server/internal/app/commands/arc_ingest.go]: FYI:
- Not exported methods don't need doc-comment - add it only if it adds valuable context 
- Things in `internal` don't need doc-comment

Here we have both - not exported (private) and in internal 🙈

**Review by chris-4chain — APPROVED** [review]

File: pkg/server/internal/app/commands/arc_ingest.go (hunk 12 above)
Comment (cain4chain) [reply pkg/server/internal/app/commands/arc_ingest.go]: Ok thanks will update.

**Review by mgosek-4chain — APPROVED** [review]
